src/rgw/rgw_client_io.h

```diff
@@ -4,83 +4,233 @@
 #ifndef CEPH_RGW_CLIENT_IO_H
 #define CEPH_RGW_CLIENT_IO_H
 
+#include <exception>
+#include <string>
 #include <streambuf>
 #include <istream>
 #include <stdlib.h>
+#include <system_error>
 
-#include "include/types.h"
+#include <boost/utility/string_ref.hpp>
 
+#include "include/types.h"
 #include "rgw_common.h"
 
-class RGWClientIO {
-  bool _account;
+namespace rgw {
+namespace io {
 
+class BasicClient {
 protected:
-  RGWEnv env;
-
   virtual void init_env(CephContext *cct) = 0;
 
 public:
-  virtual ~RGWClientIO() {}
-  RGWClientIO() : _account(false) {}
+  virtual ~BasicClient() {}
 
   void init(CephContext *cct);
-  RGWEnv& get_env() { return env; }
+  virtual RGWEnv& get_env() noexcept = 0;
+  virtual int complete_request() = 0;
+}; /* rgw::io::Client */
+
+
+class Accounter {
+public:
+  virtual ~Accounter() {}
+
+  virtual void set_account(bool enabled) = 0;
```

Comment:
this part of the interface seems unnecessary to me.. i only see one caller, at the end of `end_header()` in rgw_rest.cc. can we just have `AccountingEngine::complete_header()` set `enabled = true;` instead?